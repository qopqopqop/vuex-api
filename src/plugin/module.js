import appendStateWithResponse from './helpers/appendStateWithResponse';
import pluginActions from './actions'
import vue from 'vue'
import axios from 'axios'
import shallowEqual from './utils'

const { CancelToken } = axios

const state = {}

// used for request cancelation
const previousStateRequest = { cancel: null, url: null, params: {}, state: 'finished' }

const actions = {
  [pluginActions.request]: ({ commit, dispatch }, { requestConfig, ...otherConfig }) => {
    const { url, method, onSuccess, onError, keyPath, params } = otherConfig
    return new Promise((resolve, reject) => {
      commit(pluginActions.request, { keyPath })
      if (url === previousStateRequest.url && previousStateRequest.state === 'started' && shallowEqual(params, previousStateRequest.params)) {
        previousStateRequest.cancel()
      }
      previousStateRequest.url = url
      previousStateRequest.state = 'started'
      previousStateRequest.params = params
      axios({
        ...requestConfig,
        ...otherConfig,
        method: method || 'GET',
        url,
        cancelToken: new CancelToken(c => {
          previousStateRequest.cancel = c
        })
      }).then(resp => {
        previousStateRequest.state = 'finished'
        commit(pluginActions.success, { resp: resp, keyPath })
        if (onSuccess) {
          const { dispatchAction, executeFunction, commitAction } = onSuccess
          if (dispatchAction) {
            const { type, payload } = dispatchAction
            dispatch(type, payload)
          }
          if (commitAction) {
            const { type, payload } = commitAction
            commit(type, payload)
          }
          if (executeFunction) {
            executeFunction(resp, { commit, dispatch })
          }
        }
        resolve(resp)
      }).catch(err => {
        if (axios.isCancel(err)) {
          return
        }
        if (onError) {
          const { dispatchAction, executeFunction, commitAction } = onError
          if (dispatchAction) {
            const { type, payload } = dispatchAction
            dispatch(type, payload)
          }
          if (commitAction) {
            const { type, payload } = commitAction
            commit(type, payload)
          }
          if (executeFunction) {
            executeFunction(err, { commit, dispatch })
          }
        }

        commit(pluginActions.error, { err: err.response, keyPath })
        reject(err)
      })
    })
  },
  [pluginActions.clear]: ({ commit }, keyPath) => {
    return new Promise((resolve, reject) => {
      commit(pluginActions.clear, keyPath)
      resolve()
    })
  }
}

const mutations = {
  [pluginActions.request]: (state, { keyPath }) => {
    const responseStateObject = { status: 'loading' }
    const { stateObject, head } = appendStateWithResponse(state, keyPath, responseStateObject)
    vue.set(state, head, stateObject)
  },
  [pluginActions.success]: (state, { keyPath, resp }) => {
    delete resp.config // make this configurable
    const responseStateObject = { status: 'success', firstCallDone: true, resp }
    const { stateObject, head } = appendStateWithResponse(state, keyPath, responseStateObject)
    vue.set(state, head, stateObject)
  },
  [pluginActions.error]: (state, { keyPath, err }) => {
    const responseStateObject = { status: 'error', err, firstCallDone: true }
    const { stateObject, head } = appendStateWithResponse(state, keyPath, responseStateObject)
    vue.set(state, head, stateObject)
  },
  [pluginActions.clear]: (state, keyPath) => {
    const responseStateObject = {}
    const { stateObject, head } = appendStateWithResponse(state, keyPath, responseStateObject)
    vue.set(state, head, stateObject)
  }
}

const module = {
  state,
  actions,
  mutations
}

export default module
