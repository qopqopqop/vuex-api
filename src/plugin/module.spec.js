/**
 * Created by vouill on 10/24/17.
 */
import module from './module'
import pluginActions from './actions'
import moxios from 'moxios'

const testRequest = { url: 'testUrl', method: 'testMethod', baseURL: 'baseUrl' }

describe('Module Actions', () => {
  beforeEach(function () {
    moxios.install()
  })

  afterEach(function () {
    moxios.uninstall()
  })
  it('should process request', () => {
    const commit = jest.fn()
    module.actions[pluginActions.request]({ commit }, testRequest)
    expect(commit.mock.calls).toMatchSnapshot()
  })

  it('should process clear', () => {
    const commit = jest.fn()
    module.actions[pluginActions.clear]({ commit }, { keyPath: 'someKeypath' })
    expect(commit.mock.calls).toMatchSnapshot()
  })

  it('should process onSuccess', () => {
    moxios.wait(() => {
      const request = moxios.requests.mostRecent()
      request.respondWith({
        status: 200,
        response: 'test'
      })
    })

    const commit = jest.fn()
    const dispatch = jest.fn()
    return module.actions[pluginActions.request]({ commit, dispatch }, { keyPath: 'someKeypath', url: 'testUrl', onSuccess: { dispatchAction: { type: 'test-type', payload: 'test-payload' } } }).then(() => {
      expect(commit.mock.calls).toMatchSnapshot()
      expect(dispatch.mock.calls).toMatchSnapshot()
    })
  })
})

describe('Module mutations', () => {
  let state = {}
  const keyPath = 'someKeyPath'
  beforeEach(() => {
    state = {}
  })
  it('should process request', () => {
    module.mutations[pluginActions.request](state, { keyPath })
    expect(state).toMatchSnapshot()
  })

  it('should process success', () => {
    module.mutations[pluginActions.success](state, { keyPath, resp: { date: 'fakeData' } })
    expect(state).toMatchSnapshot()
  })

  it('should process error', () => {
    module.mutations[pluginActions.error](state, { keyPath, err: { data: 'fakeError' } })
    expect(state).toMatchSnapshot()
  })

  it('should process clear', () => {
    module.mutations[pluginActions.clear](state, keyPath)
    expect(state).toMatchSnapshot()
  })

  it('should process request -> success', () => {
    module.mutations[pluginActions.clear](state, keyPath)
    module.mutations[pluginActions.success](state, { keyPath, resp: { date: 'fakeData' } })
    expect(state).toMatchSnapshot()
  })

  it('should process request -> error', () => {
    module.mutations[pluginActions.clear](state, keyPath)
    expect(state).toMatchSnapshot()
    module.mutations[pluginActions.error](state, { keyPath, err: { data: 'fakeError' } })
    expect(state).toMatchSnapshot()
  })
})
