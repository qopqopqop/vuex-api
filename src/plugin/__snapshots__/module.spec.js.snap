// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module Actions should process clear 1`] = `
Array [
  Array [
    "Clear_Api_Call",
    Object {
      "keyPath": "someKeypath",
    },
  ],
]
`;

exports[`Module Actions should process onSuccess 1`] = `
Array [
  Array [
    "Request_Api_Call",
    Object {
      "keyPath": "someKeypath",
    },
  ],
  Array [
    "Success_Api_Call",
    Object {
      "keyPath": "someKeypath",
      "resp": Response {
        "code": undefined,
        "config": Object {
          "adapter": [Function],
          "cancelToken": CancelToken {
            "promise": Promise {},
          },
          "data": undefined,
          "headers": Object {
            "Accept": "application/json, text/plain, */*",
          },
          "keyPath": "someKeypath",
          "maxContentLength": -1,
          "method": "get",
          "onSuccess": Object {
            "dispatchAction": Object {
              "payload": "test-payload",
              "type": "test-type",
            },
          },
          "timeout": 0,
          "transformRequest": Object {
            "0": [Function],
          },
          "transformResponse": Object {
            "0": [Function],
          },
          "url": "testUrl",
          "validateStatus": [Function],
          "xsrfCookieName": "XSRF-TOKEN",
          "xsrfHeaderName": "X-XSRF-TOKEN",
        },
        "data": "test",
        "headers": undefined,
        "request": Request {
          "config": Object {
            "adapter": [Function],
            "cancelToken": CancelToken {
              "promise": Promise {},
            },
            "data": undefined,
            "headers": Object {
              "Accept": "application/json, text/plain, */*",
            },
            "keyPath": "someKeypath",
            "maxContentLength": -1,
            "method": "get",
            "onSuccess": Object {
              "dispatchAction": Object {
                "payload": "test-payload",
                "type": "test-type",
              },
            },
            "timeout": 0,
            "transformRequest": Object {
              "0": [Function],
            },
            "transformResponse": Object {
              "0": [Function],
            },
            "url": "testUrl",
            "validateStatus": [Function],
            "xsrfCookieName": "XSRF-TOKEN",
            "xsrfHeaderName": "X-XSRF-TOKEN",
          },
          "headers": Object {
            "Accept": "application/json, text/plain, */*",
          },
          "reject": [Function],
          "resolve": [Function],
          "responseType": undefined,
          "timeout": 0,
          "url": "testUrl",
          "withCredentials": false,
        },
        "status": 200,
        "statusText": undefined,
      },
    },
  ],
]
`;

exports[`Module Actions should process onSuccess 2`] = `
Array [
  Array [
    "test-type",
    "test-payload",
  ],
]
`;

exports[`Module Actions should process request 1`] = `
Array [
  Array [
    "Request_Api_Call",
    Object {
      "keyPath": undefined,
    },
  ],
]
`;

exports[`Module mutations should process clear 1`] = `
Object {
  "someKeyPath": Object {},
}
`;

exports[`Module mutations should process error 1`] = `
Object {
  "someKeyPath": Object {
    "err": Object {
      "data": "fakeError",
    },
    "status": "error",
  },
}
`;

exports[`Module mutations should process request -> error 1`] = `
Object {
  "someKeyPath": Object {},
}
`;

exports[`Module mutations should process request -> error 2`] = `
Object {
  "someKeyPath": Object {
    "err": Object {
      "data": "fakeError",
    },
    "status": "error",
  },
}
`;

exports[`Module mutations should process request -> success 1`] = `
Object {
  "someKeyPath": Object {
    "firstCallDone": true,
    "resp": Object {
      "date": "fakeData",
    },
    "status": "success",
  },
}
`;

exports[`Module mutations should process request 1`] = `
Object {
  "someKeyPath": Object {
    "status": "loading",
  },
}
`;

exports[`Module mutations should process success 1`] = `
Object {
  "someKeyPath": Object {
    "firstCallDone": true,
    "resp": Object {
      "date": "fakeData",
    },
    "status": "success",
  },
}
`;
